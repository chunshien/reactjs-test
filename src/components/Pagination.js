import React from 'react';

class Pagination extends React.Component  {
    constructor(props) {
        super(props);  
        this.previousClick = this.previousClick.bind(this);
        this.nextClick = this.nextClick.bind(this);
        this.pageClick = this.pageClick.bind(this);          
    }
    previousClick(){
        if(this.props.activePage > 1){
            this.props.onChanged(this.props.activePage - 1)
        } 
    }
    nextClick(){
        const lastPage = Math.ceil(this.props.totalItem / this.props.itemCountPerPage);
        if(this.props.activePage < lastPage){
            this.props.onChanged(this.props.activePage + 1)
        }   
    }
    pageClick(index){        
        this.props.onChanged(index)
    }
    render() {
        const {activePage, totalItem, itemCountPerPage} = this.props;
        const lastPage = Math.ceil(totalItem / itemCountPerPage);
        console.log(lastPage)
        const firstIndex = activePage > 1 ? 
                    (activePage < lastPage ? activePage - 1 : activePage - 2)
                    : 1;
        const secondIndex = activePage > 1 ? 
                    (activePage < lastPage ? activePage : activePage - 1) 
                    : 2;
        const lastIndex = activePage < lastPage ? 
                    (activePage > 1 ? activePage + 1 : activePage + 2) 
                    : lastPage;

        console.log(firstIndex,secondIndex,lastIndex)
        return (
            <div className="pagination">
                <span 
                    onClick={this.previousClick} 
                    className={activePage === 1 ? 'disabled-page-navigation' : 'enabled-page-navigation'}>
                    {'< Previous page'}
                </span>
                <span 
                    onClick={()=>this.pageClick(firstIndex)} 
                    className={activePage === 1 ? 'active-page' : ''}>
                    {firstIndex}
                </span>
                {totalItem / itemCountPerPage >= 2 &&
                    <span 
                        onClick={()=>this.pageClick(secondIndex)} 
                        className={activePage > 1 && activePage < lastPage ? 'active-page' : ''}>
                        {secondIndex}
                    </span>
                }
                {totalItem / itemCountPerPage >= 3 &&
                    <span 
                        onClick={()=>this.pageClick(lastIndex)} 
                        className={activePage === lastPage ? 'active-page' : ''}>
                        {lastIndex}
                    </span>
                }
                <span onClick={this.nextClick} className={activePage !== lastPage ? 'enabled-page-navigation' : 'disabled-page-navigation'}>
                    {'Next page >'}
                </span>
                <style jsx ='true'>{`
                    .pagination {  
                        margin-top: 30px;
                        text-align: right;  
                        -khtml-user-select: none;
                        -moz-user-select: none;
                        -ms-user-select: none;
                        user-select: none;
                        -webkit-touch-callout: none;
                        -webkit-user-select: none;                      
                    }  
                    .pagination span{
                        padding: 5px 10px;
                        min-width: 20px;
                        cursor: pointer;
                    }                    
                    .active-page{
                        background-color: white;
                        border-bottom: 5px solid black;
                    }
                    .enabled-page-navigation{
                        color: black;
                    }
                    .disabled-page-navigation{
                        color: #dddddd;
                        cursor: not-allowed !important;
                    }
                `}</style>
            </div>
        );
    }
}
    
export default Pagination;