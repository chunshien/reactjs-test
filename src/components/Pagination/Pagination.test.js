import React from 'react';
import Pagination from './Pagination.js';
import renderer from 'react-test-renderer';

describe('Component - Pagination Test', () => {
    it('renders Pagination without crashing', () => {
        const component = renderer.create(
            <Pagination/>    
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot()        
    });

    it('Test Pagination - page 1', () => {
        const component = renderer.create(
            <Pagination 
                totalItem={20}
                itemCountPerPage={8}
                activePage={1}                            
            />    
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot()  
        
        //test previous page
        expect(tree.children[0].children[0]).toEqual('< Previous page')
        expect(tree.children[0].props.className).toEqual('disabled-page-navigation')
        
        //test first index page
        expect(tree.children[1].children[0]).toEqual('1')
        expect(tree.children[1].props.className).toEqual('active-page')
        
        //test second index page        
        expect(tree.children[2].children[0]).toEqual('2')
        expect(tree.children[2].props.className).toEqual('')
        
        //test third index page                
        expect(tree.children[3].children[0]).toEqual('3')
        expect(tree.children[3].props.className).toEqual('')

        //test next page                
        expect(tree.children[4].children[0]).toEqual('Next page >')
        expect(tree.children[4].props.className).toEqual('enabled-page-navigation')       
    });

    it('Test Pagination - page 2', () => {
        const component = renderer.create(
            <Pagination 
                totalItem={20}
                itemCountPerPage={8}
                activePage={2}                            
            />    
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot()  
        
        //test previous page
        expect(tree.children[0].children[0]).toEqual('< Previous page')
        expect(tree.children[0].props.className).toEqual('enabled-page-navigation')
        
        //test first index page
        expect(tree.children[1].children[0]).toEqual('1')
        expect(tree.children[1].props.className).toEqual('')
        
        //test second index page        
        expect(tree.children[2].children[0]).toEqual('2')
        expect(tree.children[2].props.className).toEqual('active-page')
        
        //test third index page                
        expect(tree.children[3].children[0]).toEqual('3')
        expect(tree.children[3].props.className).toEqual('')

        //test next page                
        expect(tree.children[4].children[0]).toEqual('Next page >')
        expect(tree.children[4].props.className).toEqual('enabled-page-navigation')       
    });

    it('Test Pagination - page 3', () => {
        const component = renderer.create(
            <Pagination 
                totalItem={20}
                itemCountPerPage={8}
                activePage={3}                            
            />    
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot()  
        
        //test previous page
        expect(tree.children[0].children[0]).toEqual('< Previous page')
        expect(tree.children[0].props.className).toEqual('enabled-page-navigation')
        
        //test first index page
        expect(tree.children[1].children[0]).toEqual('1')
        expect(tree.children[1].props.className).toEqual('')
        
        //test second index page        
        expect(tree.children[2].children[0]).toEqual('2')
        expect(tree.children[2].props.className).toEqual('')
        
        //test third index page                
        expect(tree.children[3].children[0]).toEqual('3')
        expect(tree.children[3].props.className).toEqual('active-page')

        //test next page                
        expect(tree.children[4].children[0]).toEqual('Next page >')
        expect(tree.children[4].props.className).toEqual('disabled-page-navigation')       
    });
})