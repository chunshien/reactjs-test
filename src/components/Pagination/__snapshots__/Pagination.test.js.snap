// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component - Pagination Test Test Pagination - page 1 1`] = `
<div
  className="pagination"
>
  <span
    className="disabled-page-navigation"
    onClick={[Function]}
  >
    &lt; Previous page
  </span>
  <span
    className="active-page"
    onClick={[Function]}
  >
    1
  </span>
  <span
    className=""
    onClick={[Function]}
  >
    2
  </span>
  <span
    className=""
    onClick={[Function]}
  >
    3
  </span>
  <span
    className="enabled-page-navigation"
    onClick={[Function]}
  >
    Next page &gt;
  </span>
  <style
    jsx="true"
  >
    
                    .pagination {  
                        margin-top: 30px;
                        text-align: right;  
                        -khtml-user-select: none;
                        -moz-user-select: none;
                        -ms-user-select: none;
                        user-select: none;
                        -webkit-touch-callout: none;
                        -webkit-user-select: none;                      
                    }  
                    .pagination span{
                        padding: 5px 10px;
                        min-width: 20px;
                        cursor: pointer;
                    }                    
                    .active-page{
                        background-color: white;
                        border-bottom: 5px solid black;
                    }
                    .enabled-page-navigation{
                        color: black;
                    }
                    .disabled-page-navigation{
                        color: #dddddd;
                        cursor: not-allowed !important;
                    }
                
  </style>
</div>
`;

exports[`Component - Pagination Test Test Pagination - page 2 1`] = `
<div
  className="pagination"
>
  <span
    className="enabled-page-navigation"
    onClick={[Function]}
  >
    &lt; Previous page
  </span>
  <span
    className=""
    onClick={[Function]}
  >
    1
  </span>
  <span
    className="active-page"
    onClick={[Function]}
  >
    2
  </span>
  <span
    className=""
    onClick={[Function]}
  >
    3
  </span>
  <span
    className="enabled-page-navigation"
    onClick={[Function]}
  >
    Next page &gt;
  </span>
  <style
    jsx="true"
  >
    
                    .pagination {  
                        margin-top: 30px;
                        text-align: right;  
                        -khtml-user-select: none;
                        -moz-user-select: none;
                        -ms-user-select: none;
                        user-select: none;
                        -webkit-touch-callout: none;
                        -webkit-user-select: none;                      
                    }  
                    .pagination span{
                        padding: 5px 10px;
                        min-width: 20px;
                        cursor: pointer;
                    }                    
                    .active-page{
                        background-color: white;
                        border-bottom: 5px solid black;
                    }
                    .enabled-page-navigation{
                        color: black;
                    }
                    .disabled-page-navigation{
                        color: #dddddd;
                        cursor: not-allowed !important;
                    }
                
  </style>
</div>
`;

exports[`Component - Pagination Test Test Pagination - page 3 1`] = `
<div
  className="pagination"
>
  <span
    className="enabled-page-navigation"
    onClick={[Function]}
  >
    &lt; Previous page
  </span>
  <span
    className=""
    onClick={[Function]}
  >
    1
  </span>
  <span
    className=""
    onClick={[Function]}
  >
    2
  </span>
  <span
    className="active-page"
    onClick={[Function]}
  >
    3
  </span>
  <span
    className="disabled-page-navigation"
    onClick={[Function]}
  >
    Next page &gt;
  </span>
  <style
    jsx="true"
  >
    
                    .pagination {  
                        margin-top: 30px;
                        text-align: right;  
                        -khtml-user-select: none;
                        -moz-user-select: none;
                        -ms-user-select: none;
                        user-select: none;
                        -webkit-touch-callout: none;
                        -webkit-user-select: none;                      
                    }  
                    .pagination span{
                        padding: 5px 10px;
                        min-width: 20px;
                        cursor: pointer;
                    }                    
                    .active-page{
                        background-color: white;
                        border-bottom: 5px solid black;
                    }
                    .enabled-page-navigation{
                        color: black;
                    }
                    .disabled-page-navigation{
                        color: #dddddd;
                        cursor: not-allowed !important;
                    }
                
  </style>
</div>
`;

exports[`Component - Pagination Test renders Pagination without crashing 1`] = `
<div
  className="pagination"
>
  <span
    className="disabled-page-navigation"
    onClick={[Function]}
  >
    &lt; Previous page
  </span>
  <span
    className="active-page"
    onClick={[Function]}
  >
    1
  </span>
  <span
    className="enabled-page-navigation"
    onClick={[Function]}
  >
    Next page &gt;
  </span>
  <style
    jsx="true"
  >
    
                    .pagination {  
                        margin-top: 30px;
                        text-align: right;  
                        -khtml-user-select: none;
                        -moz-user-select: none;
                        -ms-user-select: none;
                        user-select: none;
                        -webkit-touch-callout: none;
                        -webkit-user-select: none;                      
                    }  
                    .pagination span{
                        padding: 5px 10px;
                        min-width: 20px;
                        cursor: pointer;
                    }                    
                    .active-page{
                        background-color: white;
                        border-bottom: 5px solid black;
                    }
                    .enabled-page-navigation{
                        color: black;
                    }
                    .disabled-page-navigation{
                        color: #dddddd;
                        cursor: not-allowed !important;
                    }
                
  </style>
</div>
`;
